_ru/tour/mixin-class-composition.md

```diff
@@ -53,26 +112,87 @@ class StringIterator(s: String) extends AbsIterator {
   }
 }
 ```
+
+{% endtab %}
+
+{% tab 'Scala 3' for=mixin-concrete-string-iterator %}
+
+```scala
+class StringIterator(s: String) extends AbsIterator:
+  type T = Char
+  private var i = 0
+  def hasNext = i < s.length
+  def next() =
+    val ch = s charAt i
+    i += 1
+    ch
+```
+
+{% endtab %}
+
+{% endtabs %}
+
 `StringIterator` принимает `String` и может быть использован для обхода по строке (например, чтоб проверить содержит ли строка определенный символ).
 
 Теперь давайте создадим трейт который тоже наследуется от `AbsIterator`.
 
+{% tabs mixin-extended-abstract-iterator class=tabs-scala-version %}
+
+{% tab 'Scala 2' for=mixin-extended-abstract-iterator %}
+
 ```scala mdoc
 trait RichIterator extends AbsIterator {
   def foreach(f: T => Unit): Unit = while (hasNext) f(next())
 }
 ```
-У этого трейта реализован метод `foreach` который постоянно вызывает переданную ему функцию `f: T => Unit` на каждом новом элементе (`next()`) до тех пор пока в итераторе содержатся элементы (`while (hasNext)`). Поскольку `RichIterator` это трейт, ему не нужно реализовывать членов абстрактного класса `AbsIterator`.
 
-Мы бы хотели объединить функциональность `StringIterator` и `RichIterator` в один класс. 
+У этого трейта реализован метод `foreach`, который постоянно вызывает переданную ему функцию `f: T => Unit`
+на каждом новом элементе (`next()`) до тех пор пока в итераторе содержатся элементы (`while (hasNext)`).
+Поскольку `RichIterator` - это трейт, ему не нужно реализовывать элементы абстрактного класса `AbsIterator`.
```

Comment:
it seems this duplicate can be deleted because it's repeated at the bottom? (162-164)

Reply:
No, it's not a dublicate because these rows are for `Scala 2` comment and 162-164 rows are for `Scala 3` comment.

The difference is that for Scala 2 we use parentheses in the while condition: `while (hasNext)`
But for Scala 3 we don't use the parentheses: `while hasNext`

Reply:
You can see it on the page https://docs.scala-lang.org/tour/mixin-class-composition.html

Please, just switch the tabs.

Reply:
Sorry, my fault. Thank you for explanation!